packages/react-data-grid/src/Canvas.tsx

```diff
@@ -54,14 +58,32 @@ type RendererProps<R> = Pick<CanvasProps<R>, 'columns' | 'cellMetaData' | 'colOv
   height: number;
   isSelected: boolean;
   scrollLeft: number;
+  isScrolling: boolean;
+  colOverscanStartIdx: number;
+  colOverscanEndIdx: number;
 };
 
-export default class Canvas<R> extends React.PureComponent<CanvasProps<R>> {
+interface CanvasState {
+  scrollTop: number;
+  scrollLeft: number;
+  scrollDirection: SCROLL_DIRECTION;
+  isScrolling: boolean;
+}
+
+export default class Canvas<R> extends React.PureComponent<CanvasProps<R>, CanvasState> {
   static displayName = 'Canvas';
 
+  readonly state: Readonly<CanvasState> = {
```

Comment:
Maybe considering to move this scroll related functions to `Grid`? Even we use `ref.current.setScrollLeft` to populate the scrolling offset, we could still split the horizontal scroll bar out of the Canvas (ag-grid, airtable are all doing like this, though google sheet and Microsoft excel online is using `<canvas/>` for rendering the cells, scroll bar is also splitted). After `viewport` is removed, I think it will be easier and clearer to see the benefit for splitting the horizontal scroll bar.
https://github.com/bvaughn/react-virtualized/issues/647#issuecomment-292397805

In `Grid.tsx`, the horizontal scroll bar only sets the `scrollLeft` to the Grid, and it can use Header ref and Canvas ref to set scroll left so everything will be in sync.

Reply:
I'm not entirely sure I understand what you're proposing, maybe let's discuss this in another PR?

Reply:
https://github.com/adazzle/react-data-grid/pull/1757/files#diff-118f6af5f1d0d8769e6f01f4d2b33362R202

I was talking about this.

If we move the `scrollLeft / scrolLTop` state in `Grid.tsx` level, we don't need to pass the `onScroll` to `Canvas` anymore, the state in `Grid.tsx` would just trigger the Header `setScrollLeft` call. Rendering the scroll bar separately can make sure `header.ref.setScrolLeft` and `canvas.ref.setScrollLeft` being called at same time, so everything will be in sync.

Reply:
But we need a component to scroll so we can call the `setScrollLeft` methods, are you suggesting we add another scroll layer on top of Canvas?

Reply:
It's technically not a scroll layer, it's just a `div` with `height: 17px` at the bottom of the current `viewport` (well, you removed it) which only handles the UI horizontal scrolling event. We know the row full width, and displaying with, so the scroll left offset will be same.